File: scripts/tools/gen_manifest.py
#!/usr/bin/env python3
import argparse
import hashlib
import json
from os import path

parser = argparse.ArgumentParser()
parser.add_argument('-a', '--appinfo', required=True)
parser.add_argument('-p', '--pkgfile', required=True)
parser.add_argument('-o', '--output', required=True)

args = parser.parse_args()

outfile = args.output

with open(args.appinfo) as f:
    appinfo = json.load(f)

with open(args.pkgfile, 'rb') as f:
    hasher = hashlib.new('sha256')
    seglen = hasher.block_size * 0x800
    segment = f.read(seglen)
    while segment:
        hasher.update(segment)
        segment = f.read(seglen)

    pkghash = hasher.hexdigest()
    pkgfile = path.basename(args.pkgfile)

with open(outfile, 'w') as f:
    json.dump({
        'id': appinfo['id'],
        'version': appinfo['version'],
        'type': 'native',
        'title': 'Moonlight',
        'appDescription': 'Open Source NVIDIA GameStream Client',
        'iconUri': 'https://github.com/mariotaku/moonlight-tv/raw/main/deploy/webos/icon.png',
        'sourceUrl': 'https://github.com/mariotaku/moonlight-tv',
        'rootRequired': False,
        'ipkUrl': pkgfile,
        'ipkHash': {
            'sha256': pkghash
        }
    }, f)

File: scripts/webos/gen_manifest.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

const {ArgumentParser} = require("argparse");
const parser = new ArgumentParser()

parser.add_argument('-a', '--appinfo', {required: true})
parser.add_argument('-p', '--pkgfile', {required: true})
parser.add_argument('-o', '--output', {required: true})
parser.add_argument('-i', '--icon', {required: true})
parser.add_argument('-l', '--link', {required: true})

const args = parser.parse_args()

const outfile = args.output

const appinfo = JSON.parse(fs.readFileSync(args.appinfo, {encoding: 'utf-8'}));

const pkghash = crypto.createHash('sha256').update(fs.readFileSync(args.pkgfile)).digest('hex')
const pkgfile = path.basename(args.pkgfile);

fs.writeFileSync(outfile, JSON.stringify({
    'id': appinfo.id,
    'version': appinfo.version,
    'type': appinfo.type,
    'title': appinfo.title,
    'appDescription': appinfo.appDescription,
    'iconUri': args.icon,
    'sourceUrl': args.link,
    'rootRequired': false,
    'ipkUrl': pkgfile,
    'ipkHash': {
        'sha256': pkghash
    }
}));